File: src/main/java/com/inspiretmstech/supavault/utils/secrets/shims/UbuntuSecretsShim.java
package com.inspiretmstech.supavault.utils.secrets.shims;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Optional;

public class UbuntuSecretsShim implements SecretsShim {

    private static final String ACCOUNT = "supavault@local";
    private static final String CLIENT_SECRET_KEY = "com.inspiretmstech.supavault";

    @Override
    public Optional<String> getSecret() {
        ProcessBuilder builder = new ProcessBuilder(
                "secret-tool",
                "lookup",
                "account", ACCOUNT, // account identifier
                "service", CLIENT_SECRET_KEY // unique service name
        );

        Process process;
        String secret;
        try {
            process = builder.start();
            BufferedReader reader = new BufferedReader(new InputStreamReader(process.getInputStream()));
            secret = reader.readLine();
            process.waitFor();
        } catch (IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }

        return Optional.ofNullable(secret);
    }

    @Override
    public void setSecret(String secret) {
        ProcessBuilder builder = new ProcessBuilder(
                "secret-tool",
                "store",
                "--label=Supavault Secret",
                "account", ACCOUNT, // account identifier
                "service", CLIENT_SECRET_KEY // unique service name
        );

        Process process;
        try {
            process = builder.start();
            process.getOutputStream().write(secret.getBytes()); // Write the secret to the process
            process.getOutputStream().close();
            process.waitFor();
        } catch (IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }

        if (process.exitValue() != 0) {
            throw new RuntimeException("Failed to store secret in Gnome Keyring");
        }
    }

    @Override
    public void removeSecret() {
        ProcessBuilder builder = new ProcessBuilder(
                "secret-tool",
                "clear",
                "account", ACCOUNT, // account identifier
                "service", CLIENT_SECRET_KEY // unique service name
        );

        Process process;
        try {
            process = builder.start();
            process.waitFor();
        } catch (IOException | InterruptedException e) {
            throw new RuntimeException(e);
        }

        if (process.exitValue() != 0) {
            throw new RuntimeException("Failed to delete secret from Gnome Keyring");
        }
    }
}

File: go/pkg/helpers/auth/secrets/shims/UbuntuSecretsShim.go
//go:build linux

package shims

import (
	"bytes"
	"encoding/json"
	"errors"
	errs "github.com/inspire-labs-tms-tech/supavault/pkg/helpers/auth/secrets/errors"
	"github.com/inspire-labs-tms-tech/supavault/pkg/models/auth"
	"log"
	"os/exec"
)

type UbuntuSecretsShim struct {
}

func (u *UbuntuSecretsShim) GetSecret() (auth.ClientCredentials, error) {

	cmd := exec.Command(
		"secret-tool", "lookup",
		"account", Account,
		"service", Service,
	)

	output, err := cmd.Output()
	if err != nil {
		return auth.ClientCredentials{}, errors.New("failed to retrieve secret from Gnome Keyring")
	}

	secret := string(bytes.TrimSpace(output))
	if secret == "" {
		return auth.ClientCredentials{}, errs.NewNotFoundError()
	}

	var client auth.ClientCredentials
	if json.Unmarshal([]byte(secret), &client) != nil {
		return auth.ClientCredentials{}, err
	}
	return client, nil
}

func (u *UbuntuSecretsShim) SetSecret(client auth.ClientCredentials) error {

	// remove existing secret, if there is one
	u.RemoveSecret()

	serialized, err := json.Marshal(client)
	if err != nil {
		log.Fatalf("error serializing secret: %v", err)
	}

	cmd := exec.Command(
		"secret-tool", "store",
		"account", Account,
		"service", Service,
	)

	// Provide the secret as input to the command.
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return errors.New("failed to create stdin pipe for storing secret")
	}

	// Start the command execution.
	if err := cmd.Start(); err != nil {
		return errors.New("failed to start secret-tool process")
	}

	// Write the secret to the command's input.
	if _, err := stdin.Write(serialized); err != nil {
		return errors.New("failed to write secret to secret-tool")
	}

	// Close the input pipe.
	if err := stdin.Close(); err != nil {
		return errors.New("failed to close stdin pipe for secret-tool")
	}

	// Wait for the command to complete.
	if err := cmd.Wait(); err != nil {
		return errors.New("failed to store secret in Gnome Keyring")
	}

	// Check the exit code of the process.
	if cmd.ProcessState.ExitCode() != 0 {
		return errors.New("secret-tool returned a non-zero exit code")
	}

	return nil
}

func (u *UbuntuSecretsShim) RemoveSecret() error {
	clearCmd := exec.Command(
		"secret-tool", "clear",
		"account", Account,
		"service", Service,
	)
	_ = clearCmd.Run()
	return nil
}
